spatialmath/capsule.go

```diff
@@ -0,0 +1,328 @@
+	if err != nil {
+		return nil, err
+	}
+	config.Type = "capsule"
+	config.R = c.radius
+	config.L = c.length
+	return json.Marshal(config)
+}
+
+// String returns a human readable string that represents the capsuleCreator.
+func (c *capsule) String() string {
+	return fmt.Sprintf("Type: Capsule, Radius: %.0f, Length: %.0f", c.radius, c.length)
+}
+
+// Label returns the label of this capsule.
+func (c *capsule) Label() string {
+	if c != nil {
+		return c.label
+	}
+	return ""
+}
+
+// Pose returns the pose of the capsule.
+func (c *capsule) Pose() Pose {
+	return c.pose
+}
+
+// AlmostEqual compares the capsule with another geometry and checks if they are equivalent.
+func (c *capsule) AlmostEqual(g Geometry) bool {
+	other, ok := g.(*capsule)
+	if !ok {
+		return false
+	}
+	return PoseAlmostEqual(c.pose, other.pose) && utils.Float64AlmostEqual(c.radius, other.radius, 1e-8)
+}
+
+// Transform premultiplies the capsule pose with a transform, allowing the capsule to be moved in space.
+func (c *capsule) Transform(toPremultiply Pose) Geometry {
+	return precalcCapsule(Compose(toPremultiply, c.pose), c.radius, c.length, c.label)
+}
+
+// ToProto converts the capsule to a Geometry proto message.
+func (c *capsule) ToProtobuf() *commonpb.Geometry {
+	return &commonpb.Geometry{
+		Center: PoseToProtobuf(c.pose),
+		GeometryType: &commonpb.Geometry_Capsule{
+			Capsule: &commonpb.Capsule{
+				RadiusMm: c.radius,
+				LengthMm: c.length,
+			},
+		},
+		Label: c.label,
+	}
+}
+
+// CollidesWith checks if the given capsule collides with the given geometry and returns true if it does.
+func (c *capsule) CollidesWith(g Geometry) (bool, error) {
+	if other, ok := g.(*box); ok {
+		return capsuleVsBoxCollision(c, other), nil
+	}
+	dist, err := c.DistanceFrom(g)
+	if err != nil {
+		return true, err
+	}
+	return dist <= CollisionBuffer, nil
+}
+
+// CollidesWith checks if the given capsule collides with the given geometry and returns true if it does.
+func (c *capsule) DistanceFrom(g Geometry) (float64, error) {
+	if other, ok := g.(*box); ok {
+		return capsuleVsBoxDistance(c, other), nil
+	}
+	if other, ok := g.(*capsule); ok {
+		return capsuleVsCapsuleDistance(c, other), nil
+	}
+	if other, ok := g.(*point); ok {
+		return capsuleVsPointDistance(c, other.position), nil
+	}
+	if other, ok := g.(*sphere); ok {
+		return capsuleVsSphereDistance(c, other), nil
+	}
+	return math.Inf(-1), newCollisionTypeUnsupportedError(c, g)
+}
+
+func (c *capsule) EncompassedBy(g Geometry) (bool, error) {
+	if other, ok := g.(*capsule); ok {
+		return capsuleInCapsule(c, other), nil
+	}
+	if other, ok := g.(*box); ok {
+		return capsuleInBox(c, other), nil
+	}
+	if other, ok := g.(*sphere); ok {
+		return capsuleInSphere(c, other), nil
+	}
+	if _, ok := g.(*point); ok {
+		return false, nil
+	}
+	return true, newCollisionTypeUnsupportedError(c, g)
+}
+
+// ToPointCloud converts a capsule geometry into []r3.Vector. This method takes one argument which determines
```

Comment:
Linter missed this because it's a method whose comment is on the interface definition.

I'll need to fix this in `Sphere` as well since that's where a lot of this was copied from.